[Shared UX] Add `<NotFound />` prompt

afgomez opened the pull request:
## Summary

Creates a shared `<NotFound />` prompt to be used when any given consumer needs to show a 404 error.

<img width="1278" alt="Screenshot 2022-11-17 at 18 06 12" src="https://user-images.githubusercontent.com/57448/202511151-a35f489b-d988-46cc-9810-5fc725e29b18.png">



<!--ONMERGE {"backportTargets":["8.6"]} ONMERGE-->

afgomez commented on packages/shared-ux/prompt/errors/src/not_found.tsx:
```diff
@@ -0,0 +1,72 @@
+    loadImage();
+  }, [colorMode]);
+
+  const icon = errorImage ? <EuiImage src={errorImage} alt="" /> : null;
```
The `alt` attribute is empty on purpose since the image is decoration, not content.

kellyemurphy commented:
Hi @afgomez we did this [404 work in the Cloud](https://github.com/elastic/cloud/pull/94718) repo, so you should be able to copy that text.

afgomez commented:
Thanks @kellyemurphy! I copied it from the EUI guidelines page. It would make sense to update the text there as well to prevent future confusion.

I'll open a ticket later in their repo.

majagrubic commented on packages/shared-ux/prompt/errors/src/not_found.tsx:
```diff
@@ -0,0 +1,76 @@
+  useEffect(() => {
+    const loadImage = async () => {
+      if (colorMode === 'LIGHT') {
+        const { default: img } = await import('./assets/404_astronaut_light.png');
```
Could be simplified into one statement?
`const image = colorMode === 'LIGHT' ? './assets/404_astronaut_light.png' : './assets/404_astronaut_dark.png`

or if we want to be really picky we could even do:
`import './assets/404_astronaut_${colorMode.toLowerCase()}.png`

majagrubic commented on packages/shared-ux/prompt/errors/src/not_found.tsx:
```diff
@@ -0,0 +1,76 @@
+/**
+ * Predefined `EuiEmptyPrompt` for 404 pages.
+ */
+export function NotFound({ actions }: NotFoundProps) {
```
Could you add storybook for this comoonent?

afgomez replied:
done!

majagrubic commented on .github/CODEOWNERS:
```diff
@@ -1042,6 +1042,7 @@ packages/shared-ux/page/no_data/impl @elastic/kibana-global-experience
 packages/shared-ux/page/no_data/mocks @elastic/kibana-global-experience
 packages/shared-ux/page/no_data/types @elastic/kibana-global-experience
 packages/shared-ux/page/solution_nav @elastic/kibana-global-experience
+packages/shared-ux/prompt/errors @elastic/shared-ux
```
It should be `@elastic/kibana-global-experience`

afgomez replied:
Good catch. This was automatically done by CI.

@elastic/kibana-operations I guess it needs to be fixed there as well.

majagrubic replied:
I see you resolved the conversation, but I don't think the owner has changed?

afgomez replied:
My git-fu failed me there for a second. Sorry about that

majagrubic reviewed (commented):
Where is this design coming from? Has anyone from design approved this?

afgomez commented:
Hi @majagrubic!

> Where is this design coming from? 

This comes from [EUI](https://elastic.github.io/eui/#/display/empty-prompt#more-types-of-empty-states). It has been [implemented in cloud](https://github.com/elastic/cloud/pull/94718) and I have a use case for it in kibana. 

> Has anyone from design approved this?

I assume so since there are already implementations (unless kibana and cloud follow different design guidelines).

majagrubic commented on packages/shared-ux/prompt/errors/src/not_found.stories.tsx:
```diff
@@ -0,0 +1,60 @@
+      <EuiPageTemplate.Section alignment="center">
+        <NotFound
+          actions={[
+            <EuiButton fill color="primary" onClick={() => {}}>
```
Nit: we can add an action here and then verify it was clicked in storybook

majagrubic reviewed (commented):
I am ok with shipping this once the `codeowners` file is updated. I will leave it without explicit approval in case @clintandrewhall still has something to add.

smith commented:
@majagrubic 

> Where is this design coming from? Has anyone from design approved this?

After initially speaking with the Shared UX team I spoke with @ryankeairns in our internal #platform-design channel on 2022-09-08 and he confirmed that we should follow the same design as we're using on cloud.

clintandrewhall commented on packages/shared-ux/prompt/errors/src/not_found.tsx:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Predefined `EuiEmptyPrompt` for 404 pages.
+ */
+export function NotFound({ actions }: NotFoundProps) {
```
I'd prefer this be named `NotFoundPrompt`

clintandrewhall replied:
nit
```suggestion
export const NotFoundPrompt = ({ actions }: NotFoundProps) => {
```

clintandrewhall commented on packages/shared-ux/prompt/errors/src/not_found.tsx:
```diff
@@ -0,0 +1,73 @@
+ */
+export function NotFound({ actions }: NotFoundProps) {
+  const { colorMode } = useEuiTheme();
+  const [errorImage, setErrorImage] = useState<string>();
```
It's unclear if we're getting an image or a path... consider a bit of specificity.

```suggestion
  const [imageSrc, setImageSrc] = useState<string>();
```

clintandrewhall commented on packages/shared-ux/prompt/errors/src/not_found.tsx:
```diff
@@ -0,0 +1,73 @@
+
+  useEffect(() => {
+    const loadImage = async () => {
+      const { default: img } = await import(
```
```suggestion
      const { default: imgSrc } = await import(
```

clintandrewhall commented on packages/shared-ux/prompt/errors/src/not_found.tsx:
```diff
@@ -0,0 +1,73 @@
+      const { default: img } = await import(
+        `./assets/404_astronaut_${colorMode.toLowerCase()}.png`
+      );
+      setErrorImage(img);
```
```suggestion
      setImageSrc(imgSrc);
```

clintandrewhall commented on packages/shared-ux/prompt/errors/src/not_found.tsx:
```diff
@@ -0,0 +1,73 @@
+    loadImage();
+  }, [colorMode]);
+
+  const icon = errorImage ? <EuiImage src={errorImage} alt="" /> : null;
```
```suggestion
  const icon = errorImage ? <EuiImage src={imageSrc} alt="" /> : null;
```

clintandrewhall commented on packages/shared-ux/prompt/errors/src/not_found.tsx:
```diff
@@ -0,0 +1,73 @@
+      body={NOT_FOUND_BODY}
+      actions={
+        actions ?? [
+          <EuiButtonEmpty iconType="arrowLeft" flush="both" onClick={goBack}>
```
nit: declare this as a const above within the functional component.  Otherwise you risk recreating the element on each render, (referential difference vs. shallow comparison).  It also makes the `goBack` const cleaner.

clintandrewhall commented on packages/shared-ux/prompt/errors/src/not_found.tsx:
```diff
@@ -0,0 +1,73 @@
+  defaultMessage: 'Go back',
+});
+
+interface NotFoundProps {
```
nit, unless this is for documentation purposes.

```suggestion
export type NotFoundProps = Pick<EuiEmptyPromptProps, 'actions'>;
```

afgomez replied:
Indeed I wrote it as an interface to have inline documentation for the prop.

<img width="585" alt="Screenshot 2022-11-29 at 10 45 02" src="https://user-images.githubusercontent.com/57448/204495148-2107cc5f-8010-4845-9bd9-518a94fa266d.png">

clintandrewhall reviewed (commented):
This is an awesome contribution, thank you!

A few requests: 

1. consider `shared-ux/prompt/not_found` as the path.
2. consider `NotFoundPrompt` as the component name.

afgomez commented:
@clintandrewhall I have addressed all feedback. I have also renamed the plugin name to reflect the new folder name.

Let me know if something else is needed

clintandrewhall reviewed (approved):
Great work.  Thanks for your efforts (and patience) on this.

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.6|**Backport failed because of merge conflicts**<br><br>You might need to backport the following PRs to 8.6:<br> - [[Fleet] Fix agent status computation to support agent v2 uppercase (#146757)](https://github.com/elastic/kibana/pull/146757)|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 145598
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.7 -->
